code/controllers/subsystem/processing/day_night.dm

```diff
@@ -0,0 +1,129 @@
+
+
+/**
+ * STATION TIMES ARE 24 HR FORMAT
+ */
+
+SUBSYSTEM_DEF(day_night)
+	name = "Day/Night Cycle"
+	wait = 6 SECONDS // Every minute, the clock moves forward 1 minutes
+	init_order = INIT_ORDER_DAY_NIGHT
+	/// The current hour
+	var/current_hour = 0
+	/// The current minute
+	var/current_minute = 0
+	/// A list of all currently loaded controllers to be handled
+	var/list/cached_controllers = list()
+	/// The amount of time we add every tick
+	var/tick_time = DAY_NIGHT_SUBSYSTEM_FIRE_INCREMENT
+	/// If it is our first time firing, we will update maps accordingly as atoms that have initialised will have overriden luminosity.
+	var/first_tick = TRUE
+
+/datum/controller/subsystem/day_night/Initialize(start_timeofday)
+	current_hour = rand(0, 23) // We set the starting station time to something random.
+	load_day_night_controller()
+	return ..()
+
+/datum/controller/subsystem/day_night/fire(resumed)
+	tick_tock(tick_time)
+	if(first_tick)
+		update_controllers(current_hour)
+		first_tick = FALSE
+/**
+ * Our internal ticky tocky time machine that will move time forward by the a set amount.
+ * Arguments:
+ * * time_to_tick - The amount of time we will be adding to our internal clock
+ */
+/datum/controller/subsystem/day_night/proc/tick_tock(time_to_tick)
+	current_minute += time_to_tick
+
+	if(current_minute >= HOUR_INCREMENT)
+		var/time_delta = (current_minute - HOUR_INCREMENT) > 0 ? current_minute - HOUR_INCREMENT : 0
+		current_minute = time_delta
+		current_hour += 1
+		if(current_hour >= MIDNIGHT_RESET)
+			current_hour = 0
+		update_controllers(current_hour) // dispite the fast run time, we only update every time it reaches an hour
+
+	SEND_SIGNAL(src, COMSIG_DAY_NIGHT_CONTROLLER_TIME_TICK, current_hour, current_minute)
+
+/**
+ * Loads the currently chosen day night controller from config, if there is one.
+ */
+/datum/controller/subsystem/day_night/proc/load_day_night_controller()
+	if(!SSmapping.config.day_night_controller)
+		return
+	var/list/z_levels = SSmapping.levels_by_trait(ZTRAIT_STATION)
+	add_controller(SSmapping.config.day_night_controller, z_levels[LAZYLEN(z_levels)])
```

Comment:
what is going on with the use of z_levels here

Reply:
This is used to determine the top most related z level as you don't want lower z levels to be affected as they're underground, and, in that case, there's no point in loading them and wasting processing power.

Reply:
are z levels always included in order?

Reply:
yep